File: 3558-find-a-safe-walk-through-a-grid/find-a-safe-walk-through-a-grid.cpp

using namespace std;

class Solution {
public:
    bool findSafeWalk(vector<vector<int>>& grid, int health) {

        int m = grid.size();
        int n = grid[0].size();

        int startHealth = health - grid[0][0];

        if (startHealth <= 0)
            return false;

        queue<pair<int,int>> q;

        vector<vector<int>> best(m, vector<int>(n, -1));

        best[0][0] = startHealth;
        q.push({0,0});

        int dir[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};

        while (!q.empty()) {

            auto [r,c] = q.front();
            q.pop();

            if (r == m-1 && c == n-1)
                return true;

            for (auto &d : dir) {

                int nr = r + d[0];
                int nc = c + d[1];

                if (nr>=0 && nc>=0 && nr<m && nc<n) {

                    int newHealth = best[r][c] - grid[nr][nc];

                    if (newHealth > 0 && newHealth > best[nr][nc]) {

                        best[nr][nc] = newHealth;
                        q.push({nr,nc});
                    }
                }
            }
        }

        return false;
    }
};

File: 3558-find-a-safe-walk-through-a-grid/find-a-safe-walk-through-a-grid.py
from collections import deque

class Solution(object):
    def findSafeWalk(self, grid, health):

        m, n = len(grid), len(grid[0])

        start_health = health - grid[0][0]

        if start_health <= 0:
            return False

        best = [[-1] * n for _ in range(m)]
        best[0][0] = start_health

        q = deque([(0, 0)])

        directions = [(1,0), (-1,0), (0,1), (0,-1)]

        while q:

            r, c = q.popleft()

            if r == m-1 and c == n-1:
                return True

            for dr, dc in directions:

                nr, nc = r + dr, c + dc

                if 0 <= nr < m and 0 <= nc < n:

                    new_health = best[r][c] - grid[nr][nc]

                    if new_health > 0 and new_health > best[nr][nc]:

                        best[nr][nc] = new_health
                        q.append((nr, nc))

        return False